File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.js
function solution(msg) {
    var answer = [];
    dic = {};
    for (let index = 1; index <= 26; index++) {
        let letter = String.fromCharCode(index + 64);
        dic[letter] = index;
    }
    let st = 0;
    let ed = 1;
    let lastVal = 27;
    
    while (ed <= msg.length) {
        let curWord = msg.slice(st, ed)
        let newWord = msg.slice(st, ed + 1)
        
        if (newWord in dic) {
            ed += 1
        } else {
            answer.push(dic[curWord])
            dic[newWord] = lastVal;
            lastVal += 1
            st = ed
            ed += 1
        }
    }
    answer.push(dic[msg.slice(st, ed)])

    return answer;
}

File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.py
def solution(msg):
    answer = []
    alphabet_dict = {}
    for index in range(1, 27):
        letter = chr(index + 64)
        alphabet_dict[letter] = index
    firstIdx = 0
    lastIdx = 1
    lastVal = 27
    while lastIdx <= len(msg):
        curWord = msg[firstIdx : lastIdx]
        newWord = msg[firstIdx : lastIdx + 1]
        if newWord in alphabet_dict:
            lastIdx += 1
        else:
            answer.append(alphabet_dict[curWord])
            alphabet_dict[newWord] = lastVal
            lastVal += 1
            firstIdx = lastIdx
            lastIdx += 1
    answer.append(alphabet_dict[msg[firstIdx:lastIdx]])
    
    # while msg:
    #     if msg[firstIdx : lastIdx] in alphabet_dict: #있는 경우
    #         answer.append(alphabet_dict[msg[firstIdx : lastIdx]])
    #         msg = msg[lastIdx:]
    #     print(answer, msg[firstIdx : lastIdx])
        # if msg[firstIdx : lastIdx+2] in alphabet_dict: #알파벳에 있는 경우
        #     answer.append(msg[firstIdx:lastIdx])
        # else:
        #     alphabet_dict[firstIdx:lastIdx] = lastVal
        #     lastVal += 1
        # lastIdx += 1
        # print(alphabet_dict)
    return answer